src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs

```diff
@@ -372,115 +374,145 @@ private bool RunTestInternalWithExecutors(IEnumerable<Tuple<Uri, string>> execut
             // Call the executor for each group of tests.
             var exceptionsHitDuringRunTests = false;
 
-            // Collecting Total Number of Adapters Discovered in Machine
+            // Collecting Total Number of Adapters Discovered in Machine.
             this.requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfAdapterDiscoveredDuringExecution, executorUriExtensionMap.Count());
 
+            bool attachedToTestHost = false;
+            var executorCache = new List<LazyExtension<ITestExecutor, ITestExecutorCapabilities>>();
             foreach (var executorUriExtensionTuple in executorUriExtensionMap)
             {
-                // Get the executor
+                // Get the executor.
                 var extensionManager = this.GetExecutorExtensionManager(executorUriExtensionTuple.Item2);
 
                 // Look up the executor.
                 var executor = extensionManager.TryGetTestExtension(executorUriExtensionTuple.Item1);
-                if (executor != null)
+
+                // Cache the executor.
+                executorCache.Add(executor);
+
+                // Check if we actually have to attach to the default test host.
+                if (!this.runContext.IsBeingDebugged || attachedToTestHost)
                 {
-                    try
+                    // We already know we should attach to the default test host, simply continue.
+                    continue;
+                }
+
+                if (!(executor.Value is ITestExecutor2) || this.ShouldAttachToTestHost(executor.Value))
+                {
+                    EqtTrace.Verbose("Attaching to default test host.");
+
+                    attachedToTestHost = true;
+                    if (!this.frameworkHandle.AttachDebuggerToProcess(Process.GetCurrentProcess().Id))
                     {
-                        if (EqtTrace.IsVerboseEnabled)
-                        {
-                            EqtTrace.Verbose(
-                                "BaseRunTests.RunTestInternalWithExecutors: Running tests for {0}",
-                                executor.Metadata.ExtensionUri);
-                        }
+                        throw new TestPlatformException("Cannot attach the debugger to the default test host.");
+                    }
+                }
+            }
 
-                        // set the active executor
-                        this.activeExecutor = executor.Value;
+            int index = -1;
+            foreach (var executorUriExtensionTuple in executorUriExtensionMap)
+            {
+                // Get the executor from cache.
+                var executor = executorCache[++index];
+                if (executor == null)
+                {
+                    // Commenting this out because of a compatibility issue with Microsoft.Dotnet.ProjectModel released on nuGet.org.
```

Comment:
What is the impact of this change? is that a code that was there and is commented out now? if not, and it is just a note for yourself then remove it

Reply:
Neither a note to self, nor a piece of code commented out now. While I was at it I refactored a bit the for loop to avoid using if-else and to decrease nesting level. I followed a slightly modified early return pattern to weed out incompatible conditions first, only in this case I continue instead of returning. You can see the piece of code you named in the previous version as well, only on the else clause.

Reply:
Then I don't get why that comment is needed, and why the commented out code needs to stay. :) If there is a piece of info that needs to be preserved for the future then please describe what the issue is and why the code should not revert to the previous state in concrete terms (optionally accompanied with links to the actual issue). Simply imagine yourself reading this comment without any extra context, and you should be able to understand what it says. I am doing just that and don't understand. If you are just moving code around, because you are refactoring to a different pattern then please delete it.

Reply:
Done.